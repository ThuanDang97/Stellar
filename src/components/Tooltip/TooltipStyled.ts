import styled from 'styled-components'

// Utils
import { pxToRem } from '@utils/theme'

// Styles
import { TooltipStyledProps } from '@self-types/components/Tooltip.props'

export const TooltipStyled = styled.div<TooltipStyledProps>`
  position: absolute;
  top: -5px;
  left: 50%;
  width: max-content;
  max-width: ${pxToRem(230)};
  padding: ${({ theme }) => pxToRem(theme.metrics.dimensions.xs)};
  border-radius: ${({ theme }) => pxToRem(theme.metrics.borderRadius.default)};

  transform: translateX(-50%) translateY(calc(-100% - 5px));

  color: ${({ theme, titleColor }) => titleColor || theme.colors.white};
  font-size: ${({ theme }) => pxToRem(theme.typography.fontSize.xxs)};
  text-align: center;
  background: ${({ theme, bgColor }) => bgColor || theme.colors.black};

  // Tooltip arrow
  &::after {
    content: '';
    position: absolute;
    bottom: -10px;
    left: 50%;

    transform: translateX(-50%);

    border: ${({ theme }) => pxToRem(theme.metrics.borderWidth.md)} solid
      transparent;
    border-top-color: ${({ theme, bgColor }) => bgColor || theme.colors.black};
  }
  @media (max-width: ${({ theme }) => pxToRem(theme.metrics.breakPoints.xs)}) {
    max-width: ${pxToRem(185)};
  }
`
