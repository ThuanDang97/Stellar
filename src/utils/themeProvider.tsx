import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
