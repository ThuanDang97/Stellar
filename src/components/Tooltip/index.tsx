import React, { useCallback, useState } from 'react'

// Props type
import { TooltipProps } from '@self-types/components/Tooltip.props'

// Components
import { ViewStyled } from '@components/styled-components'

// Styles
import { TooltipStyled } from './TooltipStyled'

export default ({ children, title, bgColor, titleColor }: TooltipProps) => {
  const [isHover, setIsHover] = useState(false)

  const hoverTarget = useCallback(() => setIsHover(true), [])
  const leaveTarget = useCallback(() => setIsHover(false), [])

  return (
    <ViewStyled
      data-testid="tooltip-wrapper"
      position="relative"
      onMouseEnter={hoverTarget}
      onMouseLeave={leaveTarget}
      display="flex"
    >
      {children}

      {isHover && (
        <TooltipStyled
          data-testid="tooltip"
          bgColor={bgColor}
          titleColor={titleColor}
        >
          {title}
        </TooltipStyled>
      )}
    </ViewStyled>
  )
}
