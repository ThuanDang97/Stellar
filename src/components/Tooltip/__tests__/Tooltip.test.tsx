import { fireEvent, screen } from '@testing-library/react'

// Components
import Tooltip from '@components/Tooltip'

// Utils
import { renderWithTheme } from '@utils/themeProvider'

describe('Tooltip Component', () => {
  it('renders tooltip unchanged', () => {
    const props = {
      children: <p>Hello</p>,
      title: 'This is the tooltip',
    }

    const { asFragment } = renderWithTheme(<Tooltip {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('displays tooltip on hover target', async () => {
    const props = {
      children: <p>Hello</p>,
      title: 'This is the tooltip',
    }

    renderWithTheme(<Tooltip {...props} />)

    const target = screen.getByTestId('tooltip-wrapper')

    fireEvent.mouseOver(target)

    expect(await screen.findByText('This is the tooltip')).toBeInTheDocument()
  })
})
