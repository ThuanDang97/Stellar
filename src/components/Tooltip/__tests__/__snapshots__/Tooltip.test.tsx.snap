// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Component renders tooltip unchanged 1`] = `
<DocumentFragment>
  <div
    class="ViewStyled-sc-rqe5sa-0 iGWxQB"
    data-testid="tooltip-wrapper"
    display="flex"
  >
    <p>
      Hello
    </p>
  </div>
</DocumentFragment>
`;
